src/parser/parse.ts

```diff
@@ -0,0 +1,204 @@
+import { UsefulToken } from '../lexer/tokens'
+import { TextField, IdentifierField, FieldValue, Field, Package } from './ipkg'
+import { tokensToString } from '../lexer/to-string'
+
+export type ParseSuccess<T> = { resultType: 'success'; value: T }
+export type ParseError = { resultType: 'error'; message: string }
+export type ParseResult<T> = ParseSuccess<T> | ParseError
+
+/**
+ * Parse the head of the package file.
+ *
+ * E.g.: `package PackageName`
+ */
+function parsePackageName(
+    tokens: Array<UsefulToken>,
+    pos: number,
+): ParseResult<{ name: string; newPos: number }> {
+    const packageIdent = tokens[pos]
+    const packageName = tokens[pos + 1]
+    if (
+        packageIdent &&
+        packageIdent.tokenType === 'ident' &&
+        packageIdent.identifier === 'package' &&
+        packageName &&
+        packageName.tokenType === 'ident'
+    ) {
+        return {
+            resultType: 'success',
+            value: { name: packageName.identifier, newPos: pos + 2 },
+        }
+    } else {
+        return {
+            resultType: 'error',
+            message: 'the ipkg file needs to start with a package description',
+        }
+    }
+}
+
+/**
+ * Return the proper field value from a token.
+ * Return `null` if it's not a valid token.
+ */
+function tokenValueToFieldValue(
+    token: UsefulToken,
+): TextField | IdentifierField | null {
+    switch (token.tokenType) {
+        case 'ident': {
+            return { fieldType: 'identifier', identifier: token.identifier }
+        }
+        case 'string-lit': {
+            return { fieldType: 'text', text: token.string }
+        }
+        default: {
+            return null
+        }
+    }
+}
+
+/**
+ * Parse a field value. The thing(s) after the "=".
+ */
+function parseFieldValue(
+    tokens: Array<UsefulToken>,
+    pos: number,
+): ParseResult<{ fieldValue: FieldValue; newPos: number }> {
+    let values = [tokens[pos]]
+    let newPos = pos + 1
+    while (
+        tokens[newPos] &&
+        tokens[newPos].tokenType === 'separator' &&
+        tokens[newPos + 1]
+    ) {
+        values.push(tokens[newPos + 1])
+        newPos += 2
+    }
+
+    // convert the parsed values to the correct `FieldValue`.
+    if (values.length === 0) {
+        return {
+            resultType: 'error',
+            message: 'expected at least one value after "="',
+        }
+    } else if (values.length === 1) {
```

Comment:
What is the reason this cases is handled separately instead of just always returning a list of values?